tests/v1/test_stats.py

```diff
@@ -0,0 +1,281 @@
+        monotonic_ts_s=running_ts,
+        new_prefill=True,
+        num_computed_tokens=3,
+        num_cached_tokens=1,
+    )
+    stats.update_from(running_update)
+    assert stats.prefill_ts_s == running_ts
+    assert stats.num_computed_tokens == 3
+    assert stats.num_cached_tokens == 1
+    assert stats.queue_duration_s == running_ts - queued_ts
+
+    # Test RUNNING again shouldn't update prefill_ts_s
+    running_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.RUNNING,
+        monotonic_ts_s=running_2_ts,
+        new_prefill=False,
+        num_computed_tokens=6,
+        num_cached_tokens=0,
+    )
+    stats.update_from(running_update)
+    assert stats.prefill_ts_s == running_ts
+    assert stats.num_computed_tokens == 6
+    # num_cached_tokens is not updated
+    assert stats.num_cached_tokens == 1
+    assert stats.last_updated_ts_s == running_2_ts
+    # prefill_start_ts_s_lst should only contain the first running/resumed
+    # running prefill update.
+    assert stats.prefill_start_ts_s_lst == [
+        running_ts,
+    ]
+
+    # Test DECODED
+    decoded_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.DECODED,
+        monotonic_ts_s=decoded_ts,
+        num_new_tokens=1,
+        token_perf_ts_ns=decoded_ts * 1e9,
+    )
+    stats.update_from(decoded_update)
+    assert stats.last_updated_ts_s == decoded_ts
+    # Since arrival
+    assert stats.first_token_latency_s == decoded_ts - arrived_ts
+    assert stats.num_output_tokens == 1
+    # Since first scheduled
+    assert stats.prefill_latency_s == 2
+
+    # Test DETOKENIZED
+    detokenized_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.DETOKENIZED,
+        monotonic_ts_s=detokenized_ts,
+    )
+    stats.update_from(detokenized_update)
+    assert stats.last_updated_ts_s == detokenized_ts
+
+    # Test another DECODE should yield correct inter token latency
+    decoded_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.DECODED,
+        monotonic_ts_s=decoded_2_ts,
+        num_new_tokens=1,
+        token_perf_ts_ns=decoded_2_ts * 1e9,
+    )
+    stats.update_from(decoded_update)
+    assert stats.output_token_latency_s_lst == [
+        decoded_2_ts - decoded_ts,
+    ]
+    assert stats.num_output_tokens == 2
+
+    # Test PREEMPTED
+    preempted_update = RequestStatsUpdate(
+        request_id=request_id,
+        type=RequestStatsUpdate.Type.PREEMPTED,
+        monotonic_ts_s=preempted_ts,
+    )
+    stats.update_from(preempted_update)
+    assert stats.last_updated_ts_s == preempted_ts
+    assert stats.preempted_ts_s_lst == [preempted_ts]
+    # States should be reset
+    assert stats.num_computed_tokens == 0
+    assert stats.num_cached_tokens == 0
+    assert stats.num_output_tokens == 0
+    assert stats.output_token_latency_s_lst == []
```

Comment:
Im not sure this is the right behavior for handling preemption. From the POV of the user, if a request is preempted they will see TPOT like:
- 10ms, 10ms, 10ms [PREEMPT], 1000ms, ...

Preemptions basically manifest to the user as big spikes in P99 TPOT. I think that the metrics we are calculating in `EngineCore` need to reflect this such that we can export it to Prometheus properly

Reply:
This comment applies to the below as well

Reply:
Great callout - I do find preemption a bit tricky. 

So right now, preemption is manifested with a huge TTFT (assuming no prefix cache hit) because we calculate TTFT with the most recent prefill time after preemption, but yeah TPOT might be low. I think the main design point here is how we take into account output tokens generated before preemption in the final stats:
1. [current PR] Disregard any previously generated output token at preemption
2. [v0] Record all output token's interval. 

I personally think 1 makes more sense:
1. The threw away tokens before preemption are opaque to users, with 2, we are actually recording more output tokens than the request eventually yields to users (e.g. 10 data points, but only 1 token yields in the end)
2. Preemption is consistently manifested by large TTFT value. with 2, if preemption happens during continuous prefilling, it caused a large TTFT and normal TPOT. But if preemption happens during decodes, it caused a large P99 TPOT, and normal TTFT.